addons/hr_holidays/models/hr_leave_allocation.py

```diff
@@ -296,12 +302,33 @@ def _get_carryover_date(self, date_from):
             carryover_date += relativedelta(years=1)
         return carryover_date
 
-    def _add_days_to_allocation(self, current_level, current_level_maximum_leave, leaves_taken, period_start, period_end):
+    def _add_days_to_allocation(self, allocation_data):
         days_to_add = self._process_accrual_plan_level(
-            current_level, period_start, self.lastcall, period_end, self.nextcall)
-        self.number_of_days += days_to_add
-        if current_level.cap_accrued_time:
-            self.number_of_days = min(self.number_of_days, current_level_maximum_leave + leaves_taken)
+            allocation_data['current_level'], allocation_data['period_start'], self.lastcall, allocation_data['period_end'], self.nextcall)
+
+        carryover_date = allocation_data.get('carryover_date', False)
+        current_level_last_date = allocation_data.get('current_level_last_date', False)
+        if self.accrued_on_carryover_date(carryover_date, allocation_data['period_end'], current_level_last_date):
+            total_accrued_days = days_to_add + self.to_accrue
+            self.number_of_days += total_accrued_days
+            self.last_accrued_days_count = total_accrued_days
+            self.to_accrue = 0
+        else:
+            self.to_accrue = days_to_add
+        if allocation_data['current_level'].cap_accrued_time:
+            self.number_of_days = min(self.number_of_days, allocation_data['current_level_maximum_leave'] + allocation_data['leaves_taken'])
+
+    def accrued_on_carryover_date(self, carryover_date, period_end, current_level_last_date):
+
+        # Accrue days on carryover date only if the carryover date coincides with the accrual date or
+        # it coincides with level_transition_date.
+        # Otherwise, don't accrue any time off days on a carryover date. All days should only be accrued on the date
+        # specified by the current accrual plan level. Instead, store the prorated portion of days in to_accrue
+        # and add it to the number_of_days on the specified date.
```

Comment:
your comment is a bit difficult to undersand :/  for example you don't explain what period_end is.